Allow users to pick, show and hide measurement units in the reports and exports.

AmadeusW opened the pull request:
Introduces `ISummaryStyle` which can be added to the `IConfig` and to participating exporters (currently `CsvExporter`). 

The style controls whether units appear in headers, whether they appear in the data and what are the units. If user did not specify units in the style, `SummaryTable` will set recommended units based on report results. 

Participating implementations of `IColumn` (currently `StatisticColumn` and two `AllocationColumn`) render their headers and values based on `ISummaryStyle`

Closes #146 - Ability to specify units / easier comparison
Closes #118 - Raw data in CSV exports
Resolves a TODO in Windows.MemoryDiagnoser


This sample program
```
            var config = ManualConfig.Create(DefaultConfig.Instance);
            config.Set(new Reports.SummaryStyle()
            {
                PrintUnitsInHeader = false,
                PrintUnitsInContent = true,
                TimeUnit = Horology.TimeUnit.Second,
                SizeUnit = Columns.SizeUnit.B
            });
            config.Add(new MemoryDiagnoser());
            config.Add(new CsvExporter(
                CsvSeparator.CurrentCulture,
                new Reports.SummaryStyle()
                {
                    PrintUnitsInHeader = true,
                    PrintUnitsInContent = false,
                    TimeUnit = Horology.TimeUnit.Millisecond,
                    SizeUnit = Columns.SizeUnit.kB
                }
            ));
            BenchmarkRunner.Run<Intro.IntroColumns>(config);
```
produces **.md** output:

 |    Method |     Mean |   StdErr |   StdDev |      Min |      Max | Allocated |
 |---------- |--------- |--------- |--------- |--------- |--------- |---------- |
 | Benchmark | 0.0305 s | 0.0066 s | 0.0209 s | 0.0101 s | 0.0507 s |    1546 B |

and **.csv** output:

![image](https://cloud.githubusercontent.com/assets/1673956/24185453/78ca803c-0e90-11e7-8127-6219155e80d3.png)

dnfclas commented:
@AmadeusW,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

AmadeusW commented on src/BenchmarkDotNet.Core/BenchmarkDotNet.Core.csproj:
```diff
@@ -8,6 +8,9 @@
     <PackageId>BenchmarkDotNet.Core</PackageId>
     <RootNamespace>BenchmarkDotNet</RootNamespace>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|net45|AnyCPU'">
```
I didn't mean to change this file, but I'm unable to compile when I revert this change. If you want to, I can revert this file before merging.

adamsitnik replied:
What kind of error were you getting? I can compile without this setting

AmadeusW replied:
Must have been one of VS moods. It works fine now without this change.

AndreyAkinshin commented:
Hey @AmadeusW, thanks for the PR! Could you rebase your commits on master? Appveyor can't check your PR because it's non-mergable.

AmadeusW commented:
Before merging, I'd like to add `QuantityType.Percentage` for the newly added `MisspredictRateColumn` so that we can evict the "%" to the header. Are there any other columns with units that I've missed?

AndreyAkinshin commented:
@AmadeusW, thanks for the PR. It was very busy week, I'm going to do a code review on this weekend.

AmadeusW commented:
I've made the change I wanted to do for the `MispredictRateColumn`, and it turns out that I didn't need to create new `QuantityType`. The PR is now ready for the review

AndreyAkinshin commented:
@AmadeusW, the build is red again.

adamsitnik commented on src/BenchmarkDotNet.Core/Columns/IColumn.cs:
```diff
@@ -16,8 +16,30 @@ public interface IColumn
+        /// <summary>
+        /// Column title formatted using the specified style.
+        /// </summary>
+        /// <param name="style"></param>
```
lack of description for params and return

adamsitnik commented on src/BenchmarkDotNet.Core/Columns/IColumn.cs:
```diff
@@ -16,8 +16,30 @@ public interface IColumn
+        /// <summary>
+        /// Value in this column formatted using the default style.
+        /// </summary>
+        /// <param name="summary"></param>
```
lack of description for params and returns

AmadeusW replied:
Can I remove `<param>` and `<returns>` leaving just the `<summary>`? There was no documentation for this method before, and I'm afraid the comments will be similar to [this](https://github.com/dotnet/BenchmarkDotNet/pull/379#discussion_r104139993)

adamsitnik replied:
sure!

adamsitnik commented on src/BenchmarkDotNet.Core/Columns/IColumn.cs:
```diff
@@ -16,8 +16,30 @@ public interface IColumn
+        /// <param name="summary"></param>
+        /// <param name="benchmark"></param>
+        /// <param name="style"></param>
+        /// <returns></returns>
```
lack of description for params and returns

adamsitnik commented on src/BenchmarkDotNet.Core/Columns/QuantityType.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace BenchmarkDotNet.Columns
+{
+    public enum QuantityType
```
perhaps `UnitType` would be a better name?

adamsitnik commented on src/BenchmarkDotNet.Core/Columns/SizeUnit.cs:
```diff
@@ -0,0 +1,46 @@
+        public static readonly SizeUnit kB = new SizeUnit("kB", "Kilobyte", BytesInKiloByte);
+        public static readonly SizeUnit MB = new SizeUnit("MB", "Megabyte", BytesInKiloByte * BytesInKiloByte);
+        public static readonly SizeUnit GB = new SizeUnit("GB", "Gigabyte", BytesInKiloByte * BytesInKiloByte * BytesInKiloByte);
+        public static readonly SizeUnit TB = new SizeUnit("TB", "Terabyte", BytesInKiloByte * BytesInKiloByte * BytesInKiloByte * BytesInKiloByte);
```
I think that `Math.Pow(BytesInKiloByte , 4)` would be better than `BytesInKiloByte * BytesInKiloByte * BytesInKiloByte * BytesInKiloByte`

AmadeusW replied:
`Math.Pow` takes `double` and returns `double`

adamsitnik commented on src/BenchmarkDotNet.Core/Columns/SizeUnit.cs:
```diff
@@ -0,0 +1,46 @@
+            return All.Last();
+        }
+
+        public static double Convert(long value, SizeUnit from, SizeUnit to) => value * (double)@from.ByteAmount / (to ?? GetBestSizeUnit(value)).ByteAmount;
```
I like the idea of putting "sizing" logic to a separate class! 👍

adamsitnik commented on src/BenchmarkDotNet.Core/Columns/StatisticColumn.cs:
```diff
@@ -47,33 +47,51 @@ public enum Priority
+            return Format(summary[benchmark].ResultStatistics, SummaryStyle.Default);
+        }
+        public string GetValue(Summary summary, Benchmark benchmark, ISummaryStyle style)
+        {
```
why did you give up on expression body (`=>`) here?

AmadeusW replied:
Probably because the column was already 90 characters long. I'll make it an expression bodied method.

adamsitnik commented on src/BenchmarkDotNet.Core/Configs/DefaultConfig.cs:
```diff
@@ -111,5 +112,10 @@ private static IDiagnoser GetDiagnoser(Assembly loadedAssembly, string typeName)
         }
+
+        public ISummaryStyle GetSummaryStyle()
+        {
```
please use expression body here as well

adamsitnik commented on src/BenchmarkDotNet.Core/Diagnosers/MemoryDiagnoser.cs:
```diff
@@ -62,13 +62,28 @@ public bool IsAvailable(Summary summary)
 
-                return results[benchmark].BytesAllocatedPerOperation.ToFormattedBytes();
+                var value = results[benchmark].BytesAllocatedPerOperation;
+                return QuantityType == QuantityType.Size ? value.ToSizeStr(style.SizeUnit, 1, style.PrintUnitsInContent) : ((double)value).ToStr();
```
IMPORTANT: the value is long for a reason. Please don't cast it to double

AmadeusW replied:
I thought exactly the same thing, but then I realized that this method is used for display only, and previously we also used `double`. Specifically, `ToFormattedBytes` did
 `double dblSByte = bytes;`
` dblSByte = bytes / (double)BytesInKiloByte;`
`return String.Format...dblSByte`

adamsitnik replied:
You are right, sorry for the false alarm.

adamsitnik commented on src/BenchmarkDotNet.Core/Diagnosers/MemoryDiagnoser.cs:
```diff
@@ -62,13 +62,28 @@ public bool IsAvailable(Summary summary)
+                else
+                {
+                    return ColumnName;
+                }
```
could be simplified ;)

adamsitnik commented on src/BenchmarkDotNet.Core/Exporters/Csv/CsvMeasurementsExporter.cs:
```diff
@@ -12,15 +13,17 @@ namespace BenchmarkDotNet.Exporters.Csv
+        public CsvMeasurementsExporter(CsvSeparator separator = CsvSeparator.CurrentCulture, ISummaryStyle style = null)
         {
             Separator = separator.ToRealSeparator();
+            Style = style ?? SummaryStyle.Default;
```
I like this approach 👍

adamsitnik commented on src/BenchmarkDotNet.Core/Exporters/Csv/CsvMeasurementsExporter.cs:
```diff
@@ -60,7 +65,7 @@ private MeasurementColumn[] GetColumns(Summary summary)
 
             var columns = new List<MeasurementColumn>(Columns.Value);
-            columns.Add(new MeasurementColumn("Gen_0", (_, report, __) => report.GcStats.Gen0Collections.ToString()));
+            columns.Add(new MeasurementColumn("Gen_0 XX", (_, report, __) => report.GcStats.Gen0Collections.ToString()));
```
why did you put the `XX` here?

AmadeusW replied:
Good catch!! I forgot to remove this.

adamsitnik commented on src/BenchmarkDotNet.Core/Exporters/Csv/CsvMeasurementsExporter.cs:
```diff
@@ -87,7 +92,7 @@ private static MeasurementColumn[] BuildColumns()
 
             // Measurements
-            columns.Add(new MeasurementColumn("Measurement_LaunchIndex", (summary, report, m) => m.LaunchIndex.ToString()));
+            columns.Add(new MeasurementColumn("Measurement_LaunchIndex YY", (summary, report, m) => m.LaunchIndex.ToString()));
```
why did you put the `YY` here?

AmadeusW replied:
Good catch!! I forgot to remove this.

adamsitnik commented on src/BenchmarkDotNet.Core/Extensions/CommonExtensions.cs:
```diff
@@ -20,24 +17,34 @@ public static List<T> ToSortedList<T>(this IEnumerable<T> values)
         }
 
-        internal static string ToFormattedBytes(this long bytes)
+        public static string ToSizeStr(this long value, SizeUnit unit = null, int unitNameWidth = 1, bool showUnit = true)
```
perhaps this method should also be put to `SizeUnit`?

AmadeusW replied:
We could do it, and I would rather use less extension methods, but then we'd split the "to string" functionality into `ToStr` in `CommonExtensions`, `SizeUnit` and `TimeUnit`. Now they are at least all in the same location. What do you think?

AmadeusW replied:
Leaving this as is for now.

adamsitnik commented on src/BenchmarkDotNet.Core/Reports/ISummaryStyle.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    public interface ISummaryStyle
+    {
+        bool PrintUnitsInHeader { get; set; }
```
this interface should expose only readonly properties (please remove setters)

adamsitnik commented on src/BenchmarkDotNet.Diagnostics.Windows/PmcDiagnoser.cs:
```diff
@@ -231,6 +234,9 @@ public MisspredictRateColumn(Dictionary<Benchmark, PmcStats> results)
             public ColumnCategory Category => ColumnCategory.Diagnoser;
             public int PriorityInCategory => 1;
+            public QuantityType QuantityType => QuantityType.None;
+            public string GetName(ISummaryStyle style) => ColumnName;
```
should the GetName implementation handle `style.PrintUnitsInContent` to include `%` sign?

AmadeusW replied:
It should not, because inclusion of percentage sign doesn't change the magnitude of the value: `0.123 == 12.3%`
We worry about this in case of units of measure , because when we remove the unit from `0.123kB` or `123B`, we need to include it in the header for the value to be meaningful.

In other words, `GetValue` of `PmcDiagnoser` renders either `0.123` or `12.3%`, but never `12.3` which is the only case that would warrant `[%]` in the column header.

Do you think this warrants a comment in the code?

adamsitnik replied:
ok. thanks for the explanation!

AndreyAkinshin commented:
It seems that the "Visual Studio 2017 RC" appveyor image is not available anymore. @AmadeusW, could you update your `appveyor.yml`? (Or you can just rebase your branch on master because @adamsitnik already fixed it there.)

AndreyAkinshin commented on src/BenchmarkDotNet.Core/Columns/SizeUnit.cs:
```diff
@@ -0,0 +1,46 @@
+        public static readonly SizeUnit TB = new SizeUnit("TB", "Terabyte", BytesInKiloByte * BytesInKiloByte * BytesInKiloByte * BytesInKiloByte);
+        public static readonly SizeUnit[] All = { B, kB, MB, GB, TB };
+
+        public static SizeUnit GetBestSizeUnit(params long[] values)
```
Could we introduce one `GetBestUnit` util method for both `TimeUnit` and `SizeUnit`?

AmadeusW replied:
Do you want to add a new interface, or just unify the method names?

AndreyAkinshin replied:
Hmm. Probably it makes sense to introduce an interface (like `IMeasurementUnit`). In the future, I want to add more unit types, it would be great to have common logic with nice API.

AndreyAkinshin replied:
By the way, I have a similar feature request: https://github.com/dotnet/BenchmarkDotNet/issues/404 I think it makes sense to add this feature right now because it's also a common logic for `IMeasurementUnit`. Could you implement it?

AmadeusW replied:
sounds good. I'll look into this

AndreyAkinshin commented on src/BenchmarkDotNet.Core/Attributes/Exporters/CsvExporterAttribute.cs:
```diff
@@ -1,10 +1,11 @@
     public class CsvExporterAttribute : ExporterConfigBaseAttribute
     {
-        public CsvExporterAttribute(CsvSeparator separator = CsvSeparator.CurrentCulture) : base(new CsvExporter(separator))
+        public CsvExporterAttribute(CsvSeparator separator = CsvSeparator.CurrentCulture) : base(new CsvExporter(separator, SummaryStyle.Default))
```
I think it's better don't specify default summary style.

AndreyAkinshin commented on src/BenchmarkDotNet.Core/Columns/QuantityType.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public enum UnitType
+    {
+        None,
```
I don't like that we use `None` value for different column types. Could we introduce some additional types which help us to understand what kind of data do we use in a column? E.g. `Number`, `String`, etc?

AmadeusW replied:
Currently we're using this to help render value in `GetValue`. I'll rename `None` to `Dimensionless` but I wouldn't add other elements until we have a use case for them

AndreyAkinshin replied:
Here is a use case: https://github.com/dotnet/BenchmarkDotNet/issues/397

AndreyAkinshin commented on src/BenchmarkDotNet.Core/Columns/SizeUnit.cs:
```diff
@@ -0,0 +1,46 @@
+
+        public const long BytesInKiloByte = 1000L;
+        public static readonly SizeUnit B = new SizeUnit("B", "Byte", 1L);
+        public static readonly SizeUnit kB = new SizeUnit("kB", "Kilobyte", BytesInKiloByte);
```
Could we use `KB` instead of `kB` (it's better to have common naming style).

AndreyAkinshin commented on src/BenchmarkDotNet.Core/Columns/StatisticColumn.cs:
```diff
@@ -74,17 +74,10 @@ public string GetValue(Summary summary, Benchmark benchmark, ISummaryStyle style
-                return ColumnName;
-            }
-        }
+        public string GetName(ISummaryStyle style) =>
```
It's better to not duplicate this method 3 times and introduce an util method.

AmadeusW replied:
That's a good idea

AndreyAkinshin commented on src/BenchmarkDotNet.Core/Extensions/CommonExtensions.cs:
```diff
@@ -60,6 +61,30 @@ public static string ToStr(this double value, string format = "0.##")
+                return column.ColumnName;
+            }
+
+            if (column.UnitType == UnitType.Size)
```
Could we use `switch` here?

AmadeusW replied:
Sure. I was just thinking how F#'s pattern matching would guarantee that all cases are handled, even when someone adds a new enum member.

AndreyAkinshin commented on src/BenchmarkDotNet.Core/Extensions/CommonExtensions.cs:
```diff
@@ -60,6 +61,30 @@ public static string ToStr(this double value, string format = "0.##")
+        public static string GetColumnTitle(this IColumn column, ISummaryStyle style)
+        {
+            if (!style.PrintUnitsInHeader)
+            {
```
In BenchmarkDotNet, we usually don't use brackets for simple one-line `if` statements.

AmadeusW replied:
good to know

AndreyAkinshin commented on src/BenchmarkDotNet.Core/Extensions/CommonExtensions.cs:
```diff
@@ -60,6 +61,30 @@ public static string ToStr(this double value, string format = "0.##")
+        /// <summary>
+        /// Gets column title formatted using the specified style
+        /// </summary>
+        public static string GetColumnTitle(this IColumn column, ISummaryStyle style)
```
Now it looks much better!

AmadeusW commented:
I think I've addressed all the comments @adamsitnik 
@AndreyAkinshin I feel that this PR is large enough and #404 deserves its own PR
I've squashed the commits in #413

adamsitnik commented:
LGTM!  Once again that you for your contribution!

@AndreyAkinshin please squash and merge if you think that this PR is OK

AmadeusW commented:
Should be good now

adamsitnik commented:
@AmadeusW thank you very much!

nietras commented:
Awesome 👍

AmadeusW commented:
Sweet. I'll come back with updated documentation so that everyone can take advantage of this!

AndreyAkinshin commented:
>  I'll come back with updated documentation so that everyone can take advantage of this!

It would be great, thanks!
